datafusion/common/src/with_span.rs

```diff
@@ -0,0 +1,96 @@
+use std::{cmp::Ordering, fmt::{self, Debug, Display}, ops::{Deref, DerefMut}};
+
+use sqlparser::tokenizer::Span;
+
+/// A wrapper type for entitites that somehow refer to a location in the source
```

Comment:
Did you consider a trait like was done in sqlparser that retrieves spans? For example

```rust
pub trait Spanned {
  /// return the span for this object, if known
  fn span(&self) -> Option<&Span>
}
```

And then we would implement that trait for various DataFusion / Arrow tyoes (like `DataType`, etc)?

Reply:
That would likely require implementing specific wrapper types like `DataTypeWithSpan` or something to attach Span information to DataTypes 

I also see the need to somehow plumb through the span information into things like coerce-types 🤔 

Though maybe now would be a good time to make a proper structure for coercing types, (`struct TypeCoercer` or something) where we could attach the spans 🤔

Reply:
> Did you consider a trait like was done in sqlparser that retrieves spans?

Yes, but to implement one such trait I'd have to put the data in `DataType`. And I think that's a bit pushing the limits of what `DataType` is meant to do. I quite like the idea of the `WithSpan` (can rename) functioning like a sidecar, to limit big changes in the code.

Maybe we could have a `Spanned` trait in datafusion as well, that is automatically implemented for all `WithSpan<T>`, and also by types that natively carry a `Span`?

> Though maybe now would be a good time to make a proper structure for coercing types, (struct TypeCoercer or something) where we could attach the spans 🤔

Not too familiar about the history of that component. But I think that's just one example, and there's probably many more cases where changes like that would be needed to accommodate the presence of `Span`s.